src/main/scala/util/gen.scala

```diff
@@ -2,56 +2,88 @@ package util
 
 import rise.elevate.rules.traversal.default
 import shine.C.primitives.imperative.CFunctionDefinition
-import shine.{DPIA, Pipe}
 import shine.OpenCL.{GlobalSize, LocalSize}
-import util.compiler.DSL.run
+import shine.{DPIA, Pipe}
+import util.compiler.DSL._
+import util.compiler.PartialCompiler
 
 object gen {
   import shine.C
 
   type Expr     = rise.core.Expr
   type Phrase   = DPIA.Phrases.Phrase[_ <: DPIA.Types.PhraseType]
-  type Module     = C.Module
+
+  private def exprToPhrase: Expr => Phrase =
+    shine.DPIA.fromRise(_)(default.RiseTraversable)
+
+  type CModule  = C.Module
   type CFunDef  = CFunctionDefinition
 
   object c {
     object function {
-      def fromExpr: Expr => Module =
+      def fromExpr: Expr => CModule =
         gen.c.function().fromExpr
 
       def asStringFromExpr: Expr => String =
         gen.c.function().asStringFromExpr
 
-      def asString: Module => String =
+      def asString: CModule => String =
         gen.functionAsString
     }
 
     case class function(name: String = "foo") {
-      def fromExpr: Expr => Module =
+      def fromExpr: Expr => CModule =
         gen.functionFromExpr(name, C.CodeGenerator())
 
       def asStringFromExpr: Expr => String =
         gen.functionAsStringFromExpr(name, C.CodeGenerator())
     }
   }
 
+  private def phraseToFunDef(name: String): Phrase => CFunDef =
+    CFunctionDefinition.fromPhrase(name)
+
+  private def funDefToFunction(gen: shine.C.CodeGenerator): CFunDef => CModule =
+    _.translateToModule(gen)
+
+  private def functionFromExpr(name: String = "foo",
+                               gen: shine.C.CodeGenerator =
+                                  shine.C.CodeGenerator()
+                              ): Expr => CModule =
+    exprToPhrase andThen
+      phraseToFunDef(name) andThen
+      funDefToFunction(gen)
+
+  private def functionAsStringFromExpr(name: String = "foo",
+                                       gen: shine.C.CodeGenerator =
+                                          shine.C.CodeGenerator()
+                                      ): Expr => String =
+    functionFromExpr(name, gen) andThen
+      C.Module.translateToString andThen
+      run(SyntaxChecker(_))
+
+  private def functionAsString: CModule => String =
+    // FIXME: SyntaxChecker disabled for host code prototype
+    C.Module.translateToString /*_ andThen
+      run(SyntaxChecker(_))*/
```

Comment:
I think that including the runtime header during syntax checking would solve this one.